File: node/src/report.js
const SEVERITY_ORDER = { fixed: 0, info: 1, warn: 2, error: 3 };
const SYMBOL = { fixed: "fixed", warn: "warn", error: "block", info: "info" };
const SYMBOL_CHECK = { fixed: "would fix", warn: "warn", error: "block", info: "info" };

function renderReport(report, write = true) {
  if (!report.findings.length) return `clean  ${report.path}`;
  const symbols = write ? SYMBOL : SYMBOL_CHECK;
  const c = report.counts;
  const bits = [];
  for (const key of ["error", "warn", "fixed"]) {
    if (c[key]) {
      const label = write || key !== "fixed" ? key : "would fix";
      bits.push(`${c[key]} ${label}`);
    }
  }
  const lines = [`${report.path}  (${bits.join(", ")})`];
  const ordered = [...report.findings].sort((a, b) => (SEVERITY_ORDER[b.severity] || 0) - (SEVERITY_ORDER[a.severity] || 0));
  for (const f of ordered) {
    let line = `  [${symbols[f.severity] || f.severity}] ${f.message} x${f.count}`;
    if (f.examples && f.examples.length) line += `  (${f.examples.join(", ")})`;
    lines.push(line);
  }
  return lines.join("\n");
}

function renderSummary(reports, write = true) {
  const totals = { fixed: 0, warn: 0, error: 0 };
  let changed = 0;
  for (const r of reports) {
    for (const key of Object.keys(totals)) totals[key] += r.counts[key] || 0;
    if (r.changed) changed += 1;
  }
  const changedLabel = write ? "changed" : "would change";
  const fixedLabel = write ? "fixed" : "would fix";
  return [
    `${reports.length} files scanned`,
    `${changed} ${changedLabel}`,
    `${totals.fixed} ${fixedLabel}`,
    `${totals.warn} warnings`,
    `${totals.error} blocking`,
  ].join("  |  ");
}

function serializeReport(report) {
  const counts = report.counts || {};
  return {
    path: report.path,
    changed: Boolean(report.changed),
    original_length: report.original_length || 0,
    cleaned_length: report.cleaned_length || 0,
    counts: {
      fixed: counts.fixed || 0,
      warn: counts.warn || 0,
      error: counts.error || 0,
      info: counts.info || 0,
    },
    findings: report.findings.map((f) => ({
      layer: f.layer,
      kind: f.kind,
      severity: f.severity,
      message: f.message,
      count: f.count,
      examples: f.examples || [],
    })),
  };
}

module.exports = { renderReport, renderSummary, serializeReport };

File: python/watermark_cleaner/report.py
from .findings import SEVERITY_ORDER

_SYMBOL = {"fixed": "fixed", "warn": "warn", "error": "block", "info": "info"}


def render_report(report, verbose=False):
    lines = []
    if not report.findings:
        return f"clean  {report.path}"
    counts = report.counts()
    summary_bits = []
    for key in ("error", "warn", "fixed"):
        if counts.get(key):
            summary_bits.append(f"{counts[key]} {key}")
    header = f"{report.path}  ({', '.join(summary_bits)})"
    lines.append(header)
    ordered = sorted(report.findings, key=lambda f: -SEVERITY_ORDER.get(f.severity, 0))
    for finding in ordered:
        tag = _SYMBOL.get(finding.severity, finding.severity)
        line = f"  [{tag}] {finding.message} x{finding.count}"
        if finding.examples:
            line += f"  ({', '.join(finding.examples)})"
        lines.append(line)
    return "\n".join(lines)


def render_summary(reports):
    totals = {"fixed": 0, "warn": 0, "error": 0}
    files_changed = 0
    files_blocked = 0
    for report in reports:
        counts = report.counts()
        for key in totals:
            totals[key] += counts.get(key, 0)
        if report.changed:
            files_changed += 1
        if report.has_blocking:
            files_blocked += 1
    parts = [
        f"{len(reports)} files scanned",
        f"{files_changed} changed",
        f"{totals['fixed']} fixed",
        f"{totals['warn']} warnings",
        f"{totals['error']} blocking",
    ]
    return "  |  ".join(parts)
